viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -4,11 +4,31 @@ import { TraceServerConfigService } from '../common/trace-server-config';
 @injectable()
 export class TraceServerServiceImpl implements TraceServerConfigService {
 
-    async startTraceServer(path: string | undefined, port: number | undefined): Promise<void> {
-        spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
+    startTraceServer(path: string | undefined, port: number | undefined): Promise<string> {
+        return new Promise<string>((resolve, reject) => {
+            const server = spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
+            server.on('error', err => {
+                reject(err.message);
+            });
+            server.on('message', () => {
+                resolve('success');
+            });
+        });
+
     }
 
-    async stopTraceServer(port: number | undefined): Promise<void> {
-        exec(`kill -9 $(lsof -t -i:${port} -sTCP:LISTEN)`);  // FIXME: Better approach to kill the server at a given port
+    stopTraceServer(port: number | undefined): Promise<string> {
+        return new Promise<string>((resolve, reject) => {
+            const killProcess = exec(`kill -9 $(lsof -t -i:${port} -sTCP:LISTEN)`);  // FIXME: Better approach to kill the server at a given port
```

Comment:
IIUC this Theia extension is able to spawn multiple trace servers, but no more than 1 per port?

If so this class should internally keep a map of "port to child_process" that you can then use to issue a kill on the process pid (like `child_process.kill()`).

Reply:
I think child_process.kill() will be more deterministic

Reply:
> IIUC this Theia extension is able to spawn multiple trace servers, but no more than 1 per port?
> Yes
> If so this class should internally keep a map of "port to child_process" that you can then use to issue a kill on the process pid (like `child_process.kill()`).

I have actually tried using process.kill(pid) but ran into an interesting behavior.

**spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);** 
The above spawn command will return the UNIX pid. The parameters provided within this spawn command is the path to the java executable which, when executed successfully will have a different process id. Hence when you execute process.kill(pid) [returned from the spawn], it won't kill the trace-server process running on the back-end. Instead, we will have to recursively kill the entire PID tree, in this case, to be able to successfully kill the trace-server as well (working on it as we speak).

There is also an issue where a user can kill any process by modifying the port parameter and running the stop trace server command. Ideally, user should only be allowed to kill the process started by them. This is also a work in progress and hence the reason why this PR is in Draft state :)

Reply:
> [...] when executed successfully will have a different process id.

Ok so if the process started by `${path}` spins another process, are stdio still plugged to the spawned process?

Also, it seems like your are implying that multiple users could start different Trace Servers on the same machine. Here each user has to explicitly set the port and make sure it's not one already in use. A solution to this UX problem is to use port 0 that usually means "let the OS pick a free one". The issue then is to know which one was picked but it's often easy to fetch, then you might want to send it up to the parent, through stdout for instance.

> Instead, we will have to recursively kill the entire PID tree [...]

Yup, see [pidtree](https://www.npmjs.com/package/pidtree) for that. You'll still need to keep track of the spawned PID as they should be the tree's root.

Reply:
> There is also an issue where a user can kill any process by modifying the port parameter and running the stop trace server command.

If the port is fetched from the preferences, you should listen for changes and act accordingly. It could be killing the server, maybe restarting it with the new port. Depends on the UX you want.

Reply:
Thanks for the feedback Paul. 
I am using [tree-kill](https://www.npmjs.com/package/tree-kill) to kill the process tree. This one has more downloads and hence this npm package looked more attractive :).